src/vision/status_quo/barbara_builds_an_async_executor.md

```diff
@@ -0,0 +1,161 @@
+# 😱 Status quo stories: Barbara builds an async executor
+
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today. 
+
+## The story
+
+Barbara wants to set priorities to the tasks spawned to the executor. However, she finds no existing async executor provides such a feature so she decided to build her own async executor.
+
+First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She could use it to build task schedulers. It is not the main part of the story. Each working thread can get a task from the scheduler and poll it in loop.
+
+But wait, what is a "task"?
+
+At first, Barbara thought it must contain the `Future` itself and the additional priority which was used by the scheduler. So she first wrote:
+
+```rust
+pub struct Task {
+    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
+    priority: u8
+}
+```
+
+And the working thread loop should run something like:
+
+```rust
+pub fn poll_task(task: Task) {
+    let waker = todo!();
+    let mut cx = Context::from_waker(&waker);
+    task.future.as_mut().poll(&mut cx);
+}
+```
+
+"How to create a waker?" Barbara asked herself. Quickly, she found the `Wake` trait. After reading the documentation, she realized the task in the scheduler should be stored in an `Arc`. And to implement `Wake`, the `Task` should also contain the sender of the scheduler. She changed the code to something like this:
```

Comment:
At this point I'm thinking about building an I/O reactor, so it might help to explain (here or above) that this is for compute tasks, not async I/O ones.

Reply:
It'd be helpful I think to say *why* the task should be in an arc. Presumably the idea is that the `Waker` will need to hold a reference to the task so that it can be scheduled; I guess it will need to also hold a reference to the queue or something like that?

Reply:
Oh, I guess you said that last part ('the sender of the scheduler")

Reply:
Oh, and I guess the `Arc` is partly because of the waker API...